src/geometry/manifold/diffeomorph.lean

```diff
@@ -39,64 +53,365 @@ variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+h.symm_image_eq_preimage s ▸ h.symm.unique_mdiff_on_image hn
+
+@[simp] lemma unique_diff_on_image (h : E ≃ₘ^n[𝕜] F) (hn : 1 ≤ n) {s : set E} :
+  unique_diff_on 𝕜 (h '' s) ↔ unique_diff_on 𝕜 s :=
+by simp only [← unique_mdiff_on_iff_unique_diff_on, unique_mdiff_on_image, hn]
+
+@[simp] lemma unique_diff_on_preimage (h : E ≃ₘ^n[𝕜] F) (hn : 1 ≤ n) {s : set F} :
+  unique_diff_on 𝕜 (h ⁻¹' s) ↔ unique_diff_on 𝕜 s :=
+h.symm_image_eq_preimage s ▸ h.symm.unique_diff_on_image hn
+
+end diffeomorph
+
+namespace continuous_linear_equiv
+
+variable (e : E ≃L[𝕜] E')
+
+/-- A continuous linear equivalence between normed spaces is a diffeomorphism. -/
+def to_diffeomorph : E ≃ₘ[𝕜] E' :=
+{ times_cont_mdiff_to_fun := e.times_cont_diff.times_cont_mdiff,
+  times_cont_mdiff_inv_fun := e.symm.times_cont_diff.times_cont_mdiff,
+  to_equiv := e.to_linear_equiv.to_equiv }
+
+@[simp] lemma coe_to_diffeomorph : ⇑e.to_diffeomorph = e := rfl
+@[simp] lemma symm_to_diffeomorph : e.to_diffeomorph.symm = e.symm.to_diffeomorph := rfl
+
+end continuous_linear_equiv
+
+namespace model_with_corners
+
+variables (I) (e : E ≃ₘ[𝕜] E')
+
+/-- Apply a diffeomorphism (e.g., a continuous linear equivalence) to the model vector space. -/
+def trans_diffeomorph (I : model_with_corners 𝕜 E H) (e : E ≃ₘ[𝕜] E') :
+  model_with_corners 𝕜 E' H :=
+{ to_local_equiv := I.to_local_equiv.trans e.to_equiv.to_local_equiv,
+  source_eq := by simp,
+  unique_diff' := by simp [range_comp e, I.unique_diff],
+  continuous_to_fun := e.continuous.comp I.continuous,
+  continuous_inv_fun := I.continuous_symm.comp e.symm.continuous }
+
+@[simp, mfld_simps] lemma coe_trans_diffeomorph : ⇑(I.trans_diffeomorph e) = e ∘ I := rfl
+@[simp, mfld_simps] lemma coe_trans_diffeomorph_symm :
+  ⇑(I.trans_diffeomorph e).symm = I.symm ∘ e.symm := rfl
+
+lemma trans_diffeomorph_range : range (I.trans_diffeomorph e) = e '' (range I) :=
+range_comp e I
+
+lemma coe_ext_chart_at_trans_diffeomorph (x : M) :
+  ⇑(ext_chart_at (I.trans_diffeomorph e) x) = e ∘ ext_chart_at I x := rfl
+
+lemma coe_ext_chart_at_trans_diffeomorph_symm (x : M) :
+  ⇑(ext_chart_at (I.trans_diffeomorph e) x).symm = (ext_chart_at I x).symm ∘ e.symm := rfl
+
+lemma ext_chart_at_trans_diffeomorph_target (x : M) :
+  (ext_chart_at (I.trans_diffeomorph e) x).target = e.symm ⁻¹' (ext_chart_at I x).target :=
+by simp only [range_comp e, e.image_eq_preimage, preimage_preimage] with mfld_simps
+
+end model_with_corners
+
+namespace diffeomorph
+
+variables (e : E ≃ₘ[𝕜] F)
+
+instance smooth_manifold_with_corners_trans_diffeomorph [smooth_manifold_with_corners I M] :
+  smooth_manifold_with_corners (I.trans_diffeomorph e) M :=
+begin
+  refine smooth_manifold_with_corners_of_times_cont_diff_on  _ _ (λ e₁ e₂ h₁ h₂, _),
+  refine e.times_cont_diff.comp_times_cont_diff_on
+    (((times_cont_diff_groupoid ⊤ I).compatible h₁ h₂).1.comp
+      e.symm.times_cont_diff.times_cont_diff_on _),
+  -- missing piece for `mfld_set_tac`
```

Comment:
You just need to register a lemma ` range (e ∘ I) = e.symm ⁻¹' (range I)` for a diffeomorph `e`, with the attribute `@[mfld_simps]`, and then you're good to go.

Reply:
Done, thanks!